Optimize mkString and addString

isaacl opened the pull request:
- Short circuit empty collections
- Switch to java Stringbuilder
- mark `mkString`/`addString` methods final, create a overridable
  function
- fix Map.addString regression and add test

isaacl commented on src/library/scala/collection/mutable/Builder.scala:
```diff
@@ -83,7 +83,7 @@ trait Builder[-A, +To] extends Growable[A] { self =>
   */
 @SerialVersionUID(3L)
-class StringBuilder(private val sb: java.lang.StringBuilder) extends AbstractSeq[Char]
+class StringBuilder(val sb: java.lang.StringBuilder) extends AbstractSeq[Char]
```
note the visibility change here

som-snytt replied:
consider `private[scala]`

dwijnand commented:
Hey @isaacl!

Thanks for the PR.

With regards to:

> This code fixes the regression; I'd like to add a test but not
> sure where to put it.

You can create a unit test `MapTest.scala` file in `test/junit/scala/collection` to assert this behaviour.

szeiger commented on src/library/scala/collection/IterableOnce.scala:
```diff
@@ -5,6 +5,7 @@ import scala.language.{higherKinds, implicitConversions}
 import scala.annotation.unchecked.uncheckedVariance
 import scala.math.{Ordering, Numeric}
 import scala.reflect.ClassTag
+import java.lang.{StringBuilder => jStringBuilder}
```
I'd prefer to call this `JStringBuilder`

szeiger commented on src/library/scala/collection/IterableOnce.scala:
```diff
@@ -983,7 +993,7 @@ trait IterableOnceOps[+A, +CC[_], +C] extends Any { this: IterableOnce[A] =>
     *  @return      the string builder `b` to which elements were appended.
     */
-  def addString(b: StringBuilder, start: String, sep: String, end: String): b.type = {
+  def addString(b: jStringBuilder, start: String, sep: String, end: String): jStringBuilder = {
```
This is part of the API and the method that needs to be overridden in subclasses. I don't like the idea of using a `j.l.StringBuilder` here unless we go with the idea of removing `s.c.m.StringBuilder` (and providing extension methods similar to `StringOps`) that was proposed on some other ticket.

isaacl replied:
I'm in favor of that. But I can make this function private and the scala sb version the override.

szeiger commented on src/library/scala/collection/IterableOnce.scala:
```diff
@@ -1018,7 +1056,7 @@ trait IterableOnceOps[+A, +CC[_], +C] extends Any { this: IterableOnce[A] =>
     *  @return      the string builder `b` to which elements were appended.
     */
-  def addString(b: StringBuilder, sep: String): StringBuilder = addString(b, "", sep, "")
+  final def addString(b: StringBuilder, sep: String): StringBuilder = addString(b, "", sep, "")
```
These final methods should probably be `@inline`, too

szeiger commented on src/library/scala/collection/StringOps.scala:
```diff
@@ -492,39 +494,52 @@ final class StringOps(private val s: String) extends AnyVal {
+      else {
+        b.append(s.charAt(0))
+        var i = 1
+        while (i < len) {
```
Did you benchmark this? From my experience with optimizing `ArrayOps.fold` I expect that this version will be slower. The original one has an extra branching operation inside the loop but it loops from 0 for which the JVM has extra optimizations.

isaacl replied:
I'll run a benchmark.

isaacl replied:
ran a jmh with java 8 and this version is significantly faster

szeiger reviewed (commented):
Take a look at https://github.com/scala/bug/issues/8680 and https://github.com/scala/collection-strawman/issues/550 regarding the history of `mkString` vs `addString`.

isaacl commented:
I think I've addressed all requested changes.

But Stream/LazyList tests are failing, due to expecting to be able to have different behavior between mkString and addString, which we should not allow.  I'm not sure how to fix this exactly.

som-snytt commented on src/library/scala/Predef.scala:
```diff
@@ -376,15 +376,15 @@ object Predef extends LowPriorityImplicits {
     def charAt(index: Int): Char                        = sequenceOfChars(index)
     def subSequence(start: Int, end: Int): CharSequence = new SeqCharSequence(sequenceOfChars.slice(start, end))
-    override def toString                               = sequenceOfChars mkString ""
+    override def toString                               = sequenceOfChars mkString
```
Probably avoid postfix notation.

Ichoran replied:
Agreed.  Please avoid postfix notation.  The dot adds zero characters anyway; it's not like it helps anything to use postfix here.

som-snytt replied:
I'm not the thought leader here, but also see https://github.com/scala/scala/pull/6559 which makes it a felony.

isaacl replied:
there's PostFix notation through the files I change in this patch. I'd prefer to leave those syntax changes for a different PR.

Ichoran replied:
That's reasonable, but you don't need to introduce _more_ postfix notation where there was only infix notation before.  Boyscout principle and all.

isaacl replied:
that's 1 less postfix notation already!

som-snytt replied:
postfix vs infix. At least something was a fix.

szeiger commented:
Since the actual implementation is in an internal method (the one with the Java StringBuilder) one option would be to add a flag that tells it whether to force or skip over lazy contents.

isaacl commented on test/junit/scala/collection/immutable/LazyListTest.scala:
```diff
@@ -340,28 +340,22 @@ class LazyListTest {
-    assertEquals(pre(2).take(1).toList.mkString + "?", str(pre(2)))
-    assertEquals(cyc(2).take(1).toList.mkString + "?", str(cyc(2)))
-    assertEquals(precyc(2,2).take(1).toList.mkString + "?", str(precyc(2,2)))
+    assertEquals("Stream(-2, ?)", pre(2).toString)
```
Stream.addString now forces, so the tests now use `toString`

toString adds some extra chars, but the results otherwise didn't change.

isaacl commented on src/library/scala/io/BufferedSource.scala:
```diff
@@ -83,17 +84,21 @@ class BufferedSource(inputStream: InputStream, bufferSize: Int)(implicit val cod
-      if (n>0) sb.appendAll(buf, 0, n)
+  /** Efficiently appends the entire remaining input. */
+  override def addString(sb: StringBuilder, start: String, sep: String, end: String): StringBuilder =
+    if (sep.length != 0) mkString.addString(sb, start, sep, end)
```
The `sep != ""` case expands to:
```scala
addString(new StringBuilder, "", "", "").toString.addString(sb, start, sep, end)
```
It uses the `""` case to convert to a String, and then passes off to StringOps for interspersing sep.

Ichoran commented on src/library/scala/collection/mutable/ArraySeq.scala:
```diff
@@ -184,6 +184,27 @@ object ArraySeq extends StrictOptimizedClassTagSeqFactory[ArraySeq] { self =>
+      val len = array.length
+      jsb.append(start)
+      if (len > 0) {
+        if (sep.isEmpty) jsb.append(array)
```
This doesn't look right!  `array.toString` doesn't produce the concatenated `toString` of its components.

isaacl replied:
This is using append(<Array[Char]>)

Ichoran replied:
Oh, I missed the context.  Whoops!

Ichoran commented on src/library/scala/collection/mutable/ArraySeq.scala:
```diff
@@ -184,6 +184,27 @@ object ArraySeq extends StrictOptimizedClassTagSeqFactory[ArraySeq] { self =>
+      if (len > 0) {
+        if (sep.isEmpty) jsb.append(array)
+        else {
+          jsb.ensureCapacity(jsb.length + len + end.length + (len - 1) * sep.length)
```
Not sure what the point of this sizing is given that the elements could be anywhere from 0 to who-knows-how-many characters long, and typically won't be length one, which means there will have to be a reallocation anyway.

isaacl replied:
This is for a char array. Elements are 1 char.

Ichoran replied:
Right.  I just looked at the diff, not the full file, so I missed the context.

Ichoran commented on src/library/scala/io/BufferedSource.scala:
```diff
@@ -83,17 +84,20 @@ class BufferedSource(inputStream: InputStream, bufferSize: Int)(implicit val cod
+      }
+      jsb.append(end)
+      sb
+    } else mkString.addString(sb, start, sep, end)
```
This is great for code reuse but not so good for memory usage.  I'd push the logic back into the body of the code.

Ichoran replied:
Alternatively, just comment that only the `sep = ""` case is efficient.  The other case is admittedly pretty weird.

isaacl replied:
I don’t think this is bad memory usage. All the chars are going into memory already.

In fact this can be quite compact with java compressed strings...  I doubt you could get better though could probably avoid some array copies.

Ichoran replied:
I guess storing the entire string twice is not as bad as it could be due to the ~1.5x expected buffer size compared to the final string size.  For such a rare case I guess it isn't worth optimizing more.

isaacl replied:
see https://github.com/scala/scala/pull/6878#discussion_r202232620

I thought it was quite clever to reuse the StringOps impl.  But I could probably adjust to use the ArrayOps impl instead for slightly better perf.  Though if we're more concerned about the memory usage, compressed Strings will let us hold more.

isaacl replied:
I'll add a comment, you're right that this version does consume more memory temporarily than before.

som-snytt commented on src/library/scala/collection/IterableOnce.scala:
```diff
@@ -1045,6 +1046,8 @@ trait IterableOnceOps[+A, +CC[_], +C] extends Any { this: IterableOnce[A] =>
   /** Displays all elements of this $coll in a string using start, end, and
     *  separator strings.
     *
+    * Delegates to addString, which can be overriden.
```
`overridden`, also on the overloads

som-snytt commented on src/library/scala/collection/IterableOnce.scala:
```diff
@@ -1045,6 +1046,8 @@ trait IterableOnceOps[+A, +CC[_], +C] extends Any { this: IterableOnce[A] =>
   /** Displays all elements of this $coll in a string using start, end, and
     *  separator strings.
```
I don't know if odd indentation matters to scaladoc? I think at one time it had a bug, maybe.

som-snytt commented on src/library/scala/collection/IterableOnce.scala:
```diff
@@ -1068,16 +1074,18 @@ trait IterableOnceOps[+A, +CC[_], +C] extends Any { this: IterableOnce[A] =>
     *  @return a string representation of this $coll. In the resulting string
     *          the string representations (w.r.t. the method `toString`)
     *          of all elements of this $coll follow each other without any
     *          separator string.
```
Worth scout's effort? `Concatenates the string representations (...) of all the elements of this $coll.`

som-snytt replied:
Also I'm not sure "Displays" is the right word. I would move the long return description to replace that first sentence, and then just `@return a string representation of this $coll`.

isaacl replied:
you seem to be reviewing the entire file for errors, instead of my PR. Feel free to make those changes in a separate PR.

som-snytt replied:
You don't need to pursue scouting issues, but I'll raise them for the record. Folks seeing mkString for the first time, and I remember my first time, will say, "Man, I love this language."

som-snytt commented on src/library/scala/collection/IterableOnce.scala:
```diff
@@ -1056,10 +1059,13 @@ trait IterableOnceOps[+A, +CC[_], +C] extends Any { this: IterableOnce[A] =>
+    * Delegates to addString, which can be overriden.
     *
     *  @param sep   the separator string.
     *  @return      a string representation of this $coll. In the resulting string
```
Needs a comma after `In the resulting string`.

isaacl replied:
I do not agree with these type of changes. I'm not changing that line.

som-snytt replied:
If you want your nice mkString to have ugly scaladoc, that's not a call I'm in a position to make. But I'm just making comments, not starting a review. Rest assured.

som-snytt commented on src/library/scala/collection/IterableOnce.scala:
```diff
@@ -1138,7 +1147,7 @@ trait IterableOnceOps[+A, +CC[_], +C] extends Any { this: IterableOnce[A] =>
     *  @return      the string builder `b` to which elements were appended.
     */
-  def addString(b: StringBuilder, sep: String): StringBuilder = addString(b, "", sep, "")
+  @inline final def addString(b: StringBuilder, sep: String): StringBuilder = addString(b, "", sep, "")
```
Since you're optimizing, has anyone asked if it's worth adding the other implementations? Arguably `xs.mkString(", ")` is the common case. 1 vs 3 appends? Or maybe strings are assumed slow and one only cares about space. I think I only `src.mkString` when slurping as in `BufferedSource`.

som-snytt replied:
Yes, I'm raising the question of whether passing the empty strings is highly optimized, or if it makes no difference worth mentioning. I don't know the answer.

isaacl replied:
not sure, I think empty strings are OK since they're consts

som-snytt commented on src/library/scala/collection/Map.scala:
```diff
@@ -281,12 +282,8 @@ trait MapOps[K, +V, +CC[_, _] <: IterableOps[_, AnyConstr, _], +C]
-  override def mkString(sep: String): String = super.mkString(sep)
-  override def mkString: String = super.mkString
+  override def addString(sb: StringBuilder, start: String, sep: String, end: String): StringBuilder =
+    iterator.map { case (k, v) => s"$k -> $v" }.addString(sb, start, sep, end)
```
This looks sillier now that the `s""` uses a `StringBuilder`. Maybe it's worth writing this out here. Or is it worth adding `addElementString` for `Map` to override?

som-snytt commented on src/library/scala/collection/StringOps.scala:
```diff
@@ -1,10 +1,12 @@
 
 import scala.collection.immutable.{ArraySeq, WrappedString}
+import scala.collection.immutable.{ArraySeq, WrappedString}
+import scala.collection.mutable.StringBuilder
```
That is just `scala.StringBuilder`. But I wondered if there were style points in using the `java.lang.StringBuilder` import and then `mutable.StringBuilder` where needed. Just looking for a cooler way besides the trite `J` prefix. The `mutable` is familiar from when distinguishing `immutable`.

som-snytt commented on src/library/scala/collection/StringOps.scala:
```diff
@@ -1,10 +1,12 @@
 package scala
 package collection
```
Could be `package scala.collection`, though the statements are not strictly identical any more.

isaacl replied:
please stop.  these comments are not helpful for the PR review.

isaacl committed:
Optimize mkString and addString

- Short circuit empty collections
- Switch to java Stringbuilder
- mark `mkString`/`addString` methods final, create a overridable
  function
- fix Map.addString regression (and add tests)
- unify addString/mkString impl for LazyList

isaacl commented:
(ready for review)

dwijnand commented on src/library/scala/collection/StringOps.scala:
```diff
@@ -1,10 +1,12 @@
 import java.util.NoSuchElementException
 
 import scala.collection.immutable.{ArraySeq, WrappedString}
+import scala.collection.immutable.{ArraySeq, WrappedString}
```
(minor) this import is duplicate

szeiger commented:
@isaacl I can't think of a good reason for `addString` and `mkString` to behave differently. IMHO the important aspect is that you can use `mkString` to force all elements without making `toString` (which would be used in log messages or debug output) do the same.

isaacl commented:
@szeiger cool, thanks for confirming that.  That's what I suspected, so the PR keeps that behavior.

szeiger commented on src/library/scala/collection/IterableOnce.scala:
```diff
@@ -1056,10 +1059,13 @@ trait IterableOnceOps[+A, +CC[_], +C] extends Any { this: IterableOnce[A] =>
     */
-  def mkString(start: String, sep: String, end: String): String =
-    addString(new StringBuilder(), start, sep, end).toString
+  @inline final def mkString(start: String, sep: String, end: String): String =
```
Do we want to inline this much code unconditionally? Usually only simple forwarders are marked `inline` (unless they are private methods that are explicitly designed to be inlined)

isaacl replied:
sg, I'll remove `inline` from this and the 3 arg mkString def in StringOps

szeiger reviewed (commented):
LGTM. I'll leave the final word to @Ichoran who introduced the difference in behavior between `mkString` and `addString`.

Ichoran commented:
I introduced a difference between `mkString` and `addString`?  That does not sound like something I ought to have done.  :+1: on removing any distinctions between them (save that zero-arg `mkString` should be able to return something static and not have to go through a `StringBuilder`).

NthPortal commented on src/library/scala/collection/mutable/Builder.scala:
```diff
@@ -137,6 +137,22 @@ class StringBuilder(private val underlying: java.lang.StringBuilder) extends Abs
+  override def toArray[B >: Char](implicit  ct: scala.reflect.ClassTag[B]) =
+    ct.runtimeClass match {
+      case java.lang.Character.TYPE => toCharArray.asInstanceOf[Array[B]]
+      case _ => super.toArray
```
~~Fixes scala/bug#11028~~

xuwei-k replied:
This is StringBuilder, not StringOps.

NthPortal replied:
oops, my mistake

dwijnand commented:
Thanks @isaacl for the PR, it's been a bit of a wild ride, this one. 😄
